hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -272,12 +277,10 @@ public boolean tryProcessWatermark(int ordinal, @Nonnull Watermark watermark) {
 
         // Note: We can't immediately emit current WM, as it could render items in buffers late.
         for (Byte wmKey : wmState.keySet()) {
-            long minimumBufferTime = minimumBufferTimes.get(wmKey);
             long lastReceivedWm = this.lastReceivedWm.getValue(wmKey);
-            long newWmTime = Math.min(minimumBufferTime, lastReceivedWm);
-            if (newWmTime > lastEmittedWm.getValue(wmKey)) {
-                pendingOutput.add(new Watermark(newWmTime, wmKey));
-                lastEmittedWm.put(wmKey, newWmTime);
+            if (lastReceivedWm > lastEmittedWm.getValue(wmKey)) {
```

Comment:
Here you immediately emit every received WM, that's wrong. You should hold back the received WM according to `wmState`. See the unit tests, you process late events.

Reply:
Ah, yeah, missed that, thanks for catching it